[16.0][IMP] stock_owner_restriction

AungKoKoLin1997 opened the pull request:
Supersede https://github.com/OCA/stock-logistics-workflow/pull/1385.
This PR does the following:

- Extend _update_available_quantity() and _update_reserved_quantity() in stock.quant
  to handle scenarios where owner_restriction should be enforced.
- Create a separate _get_owner_restriction() method to improve clarity and extensibility for other modules.
- Fix read_group() in stock.quant, which was incorrectly assigning the owner record
  instead of the owner ID in the domain.
- Add compute method for restrict_partner_id to dynamically control owner assignment
  based on picking type restrictions.
- Unlink stock.move.line records when owner_id is updated and owner_restriction is set.
- Update test cases.

@qrtl QT4207

AungKoKoLin1997 committed:
[IMP] stock_owner_restriction: Extend stock.quant and stock.move.line logic to support owner restriction

- Extend _update_available_quantity() and _update_reserved_quantity() in stock.quant
  to handle scenarios where owner_restriction should be enforced.
- Create a separate _get_owner_restriction() method to improve clarity and extensibility for other modules.
- Fix read_group() in stock.quant, which was incorrectly assigning the owner record
  instead of the owner ID in the domain.
- Add compute method for restrict_partner_id to dynamically control owner assignment
  based on picking type restrictions.
- Unlink stock.move.line records when owner_id is updated and owner_restriction is set.
- Update test cases.

kanda999 reviewed (approved):
LGTM: This update has been used in production for a year.

yostashiro commented on stock_owner_restriction/models/stock_picking.py:
```diff
@@ -8,3 +8,11 @@ class StockPicking(models.Model):
+    def write(self, vals):
+        if "owner_id" in vals:
+            for pick in self:
+                owner_restriction = pick.picking_type_id.owner_restriction
+                if owner_restriction in ("unassigned_owner", "picking_partner"):
```
```suggestion
                if pick.owner_restriction in ("unassigned_owner", "picking_partner"):
```

yostashiro reviewed (commented):
Could we perhaps make the intention of adding _compute_restrict_partner_id() clearer?

yostashiro commented on stock_owner_restriction/models/stock_move.py:
```diff
@@ -1,14 +1,32 @@
 class StockMove(models.Model):
     _inherit = "stock.move"
 
+    restrict_partner_id = fields.Many2one(
+        # Add compute to override the standard 'restrict_partner_id' behavior:
+        # assign the owner only when the picking type's restriction is 'picking_partner',
+        # and clear it for other restriction types to avoid unintended ownership.
```
I think this should do. No need to explain the content of the compute method.
```suggestion
    # Add compute method to the standard field
    restrict_partner_id = fields.Many2one(
```

yostashiro commented on stock_owner_restriction/models/stock_move.py:
```diff
@@ -1,14 +1,30 @@
+    @api.depends("picking_type_id", "picking_id.owner_id", "move_dest_ids")
+    def _compute_restrict_partner_id(self):
+        for move in self:
+            if move.picking_type_id.owner_restriction == "picking_partner":
```
Please check the validity of this suggestion.
```suggestion
            if move.picking_type_id.owner_restriction in ("picking_partner", "partner_or_unassigned"):
```

AungKoKoLin1997 replied:
I think this is not valid. Even if we have move lines from an unassigned owner together with move lines from an assigned owner, the move will still be assigned to the partner. Therefore, we should continue allowing False for this restriction type.

AungKoKoLin1997 committed:
[IMP] stock_owner_restriction: complete restrict_partner_id dependencies

_compute_restrict_partner_id reads move_dest_ids.picking_id.owner_id and
picking_id.partner_id via _get_owner_for_assign(), but these were not in
@api.depends. As the field is stored, the value could go stale when the
destination picking owner or the picking partner changes.
